tool/practice.re

```diff
@@ -23,7 +53,332 @@ fn create_practice_menu() -> Ui {
+            label: Text { text: "Save Custom Practice State" },
+            onclick: fn(label: Text) {
+                let practice_list = Tas::list_practice_states();
+                enter_ui(Ui::new_filechooser("Save Custom Practice State", practice_list, fn(input: string) {
+                    if input.len_utf8() == 0 {
+                        return;
+                    }
+                    let level_state = Tas::get_level_state();
+                    let loc = Tas::get_location();
+                    let rot = Tas::get_rotation();
+                    let vel = Tas::get_velocity();
+                    Tas::save_practice_state(input, Practice {
+                        name: input,
+                        cluster: level_state.level,
+                        button: level_state.buttons,
+                        location: Location { x: loc.x, y: loc.y, z: loc.z },
+                        rotation: Rotation { pitch: rot.pitch, yaw: rot.yaw, roll: rot.roll },
+                        velocity: Velocity { x: vel.x, y: vel.y, z: vel.z }
+                    });
+                    leave_ui();
+                    leave_ui();
+                }));
+            },
+        }),
+       UiElement::Button(UiButton {
+            label: Text { text: "Edit Custom Practice State" },
+            onclick: fn(label: Text) {
+                let practice_list = Tas::list_practice_states();
+                enter_ui(Ui::new_filechooser("Edit Custom Practice State", practice_list, fn(input: string) {
+                    if input.len_utf8() == 0 {
+                        return;
+                    }
+                    enter_ui(create_edit_practice_state_ui(Tas::load_practice_state(input)));
+                }));
+            },
+        }),
+        UiElement::Button(UiButton {
+            label: Text { text: "Load Custom Practice State" },
+            onclick: fn(label: Text) {
+                let practice_list = Tas::list_practice_states();
+                enter_ui(Ui::new_filechooser("Load Custom Practice State", Tas::list_practice_states(), fn(input: string) {
+                    if !practice_list.contains(input) {
+                        return;
+                    }
+                    CURRENT_PRACTICE = Tas::load_practice_state(input);
+                    add_component(PRACTICE_COMPONENT);
+                    leave_ui();
+                    leave_ui();
+                    leave_ui();
+                }));
+            },
+        }),
+        UiElement::Button(UiButton {
+            label: Text { text: "Back" },
+            onclick: fn(label: Text) { leave_ui() },
+        }),
+    );
+    Ui::new("Custom Practice:", buttons)
+}
+
+fn create_edit_practice_state_ui(mut practice: Practice) -> Ui {
+    let submit = fn() {
+        let selected = match UI_STACK.last() {
+            Option::Some(ui) => ui.selected,
+            Option::None => panic("we are currently in a UI"),
+        };
```

Comment:
This could also be changed in `map_editor.re` as there is the same lines of code present in that file, too.

Reply:
I might want to add [`Option::expect`](https://doc.rust-lang.org/std/option/enum.Option.html#method.expect) and [`Result::expect`](https://doc.rust-lang.org/std/result/enum.Result.html#method.expect) to the rebo stdlib so that we can keep the error message in the panic. Or you could do that as well if you want :)